### PR TITLE
DRIV-11: Adds support for recursive references

### PR DESCRIPTION
**Pull request by marrony**

### CODE REVIEW

File: src/Client.js [hunk 1]
```diff
@@ -95,72 +95,12 @@ Client.prototype.paginate = function(expression, params) {
 };
 
 /**
- * Issues a HTTP `GET` request via the legacy REST API.
```
Comment (erickpintor) [inline src/Client.js]: Removing this means that we should release a major version and also document it as part of the CHANGELOG.

File: src/_json.js [hunk 2]
```diff
@@ -20,7 +20,15 @@ function json_parse(_, val) {
   if (typeof val !== 'object' || val === null) {
     return val;
   } else if ('@ref' in val) {
-    return new values.Ref(val['@ref']);
+    var ref = val['@ref'];
+
+    if (!('class' in ref) && !('database' in ref))
+      return values.Native.fromName(ref['id']);
+
+    var cls = json_parse(_, ref['class']);
```
Comment (erickpintor) [inline src/_json.js]: I think it's preferable to call `parseJSON`, instead. I don't feel comfortable passing the `_` variable further.

File: src/values.js [hunk 3]
```diff
@@ -34,62 +35,54 @@ util.inherits(Value, Expr);
...
-  /**
-   * The string value of the ref.
-   *
-   * @type {string}
-   */
-  this.value = parts.join('/');
+function Ref(id, clazz, database) {
+  this.value = {};
+
+  if (id) this.value['id'] = id;
```
Comment (erickpintor) [inline src/values.js]: Shouldn't we throw an exception if no `id`?

File: src/values.js [hunk 4]
```diff
@@ -99,7 +92,9 @@ Ref.prototype.toJSON = function() {
 
 /** @ignore */
 Ref.prototype.toString = function() {
-  return this.value;
+  var cls = this.class !== undefined ? ', class=' + this.class.toString() : '';
+  var db = this.database !== undefined ? ' database = ' +  this.database.toString() : '';
```
Comment (erickpintor) [inline src/values.js]: class and database format should be equal.

File: src/query.js [hunk 5]
```diff
@@ -1053,6 +1173,14 @@ module.exports = {
   Database: Database,
   Index: Index,
   Class: Class,
+  Function: FunctionFn,
```
Comment (erickpintor) [inline src/query.js]: I almost forgot. We need to add this functions to `query.d.ts` too.

File: src/_json.js (hunk 2 above)
Comment (marrony) [reply src/_json.js]: `parseJSON` receives a string not an object

File: src/values.js [hunk 6]
```diff
@@ -99,7 +92,9 @@ Ref.prototype.toJSON = function() {
 
 /** @ignore */
 Ref.prototype.toString = function() {
-  return this.value;
+  var cls = this.class !== undefined ? ', class=' + this.class.toString() : '';
+  var db = this.database !== undefined ? ' database=' +  this.database.toString() : '';
+  return 'Ref(id=' + this.id + cls + db + ')';
```
Comment (erickpintor) [inline src/values.js]: I just realized this strings are a big part of our dashboard repl. People can right now just copy and past references' representations back into the repr and run them. I think we should keep this behaviour by making the string representation of a Ref type a valid Ref constructor.

Comment (marrony) [reply src/values.js]: this will be tricky, wouldn't be easier to the dashboard manage this?

Comment (erickpintor) [reply src/values.js]: That's a good point. I'll open a ticket for the dashboard work.

**Pull request merged**